tests/test-middle.cpp

```diff
@@ -0,0 +1,235 @@
+    }
+
+    size_t add_way(osmid_t wid, std::vector<osmid_t> const &ids)
+    {
+        using namespace osmium::builder::attr;
+        return osmium::builder::add_way(buf, _id(wid), _nodes(ids));
+    }
+
+    template <typename T>
+    T const &get(size_t pos) const
+    {
+        return buf.get<T>(pos);
+    }
+
+    template <typename T>
+    T &get(size_t pos)
+    {
+        return buf.get<T>(pos);
+    }
+
+    osmium::memory::Buffer buf{4096, osmium::memory::Buffer::auto_grow::yes};
+};
+
+void expect_location(osmium::Location loc, osmium::Node const &expected)
+{
+    CHECK(loc.lat() == Approx(expected.location().lat()));
+    CHECK(loc.lon() == Approx(expected.location().lon()));
+}
+} // namespace
+
+struct options_slim_default
+{
+    static options_t options(pg::tempdb_t const &tmpdb)
+    {
+        return testing::opt_t().slim(tmpdb);
+    }
+};
+
+struct options_slim_dense_cache : options_slim_default
+{
+    static options_t options(pg::tempdb_t const &tmpdb)
+    {
+        options_t o = options_slim_default::options(tmpdb);
+        o.alloc_chunkwise = ALLOC_DENSE;
+
+        return o;
+    }
+};
+
+struct options_ram_optimized
+{
+    static options_t options(pg::tempdb_t const &)
+    {
+        options_t o = testing::opt_t();
+        o.alloc_chunkwise = ALLOC_SPARSE | ALLOC_DENSE;
+        return o;
+    }
+};
+
+struct options_ram_flatnode
+{
+    static options_t options(pg::tempdb_t const &)
+    {
+        options_t o = testing::opt_t().flatnodes();
+        o.alloc_chunkwise = ALLOC_SPARSE | ALLOC_DENSE;
+        return o;
+    }
+};
+
+TEMPLATE_TEST_CASE("middle import", "", options_slim_default,
+                   options_slim_dense_cache, options_ram_optimized,
+                   options_ram_flatnode)
+{
+    options_t options = TestType::options(db);
+
+    auto mid = options.slim
+                   ? std::shared_ptr<middle_t>(new middle_pgsql_t(&options))
+                   : std::shared_ptr<middle_t>(new middle_ram_t(&options));
+
+    mid->start();
+
+    auto mid_q = mid->get_query_instance(mid);
+
+    test_buffer_t buffer;
+
+    SECTION("Set and retrieve a single node")
+    {
+        auto const &node = buffer.get<osmium::Node>(
+            buffer.add_node(1234, 12.3456789, 98.7654321));
+
+        // set the node
+        mid->nodes_set(node);
+        mid->flush(osmium::item_type::way);
+
+        // getting it back works only via a waylist
+        auto &nodes =
+            buffer.get<osmium::Way>(buffer.add_way(3, {1234})).nodes();
+
+        // get it back
+        REQUIRE(mid_q->nodes_get_list(&(nodes)) == nodes.size());
+        expect_location(nodes[0].location(), node);
+
+        // other nodes are not retrievable
+        auto &n2 =
+            buffer.get<osmium::Way>(buffer.add_way(3, {1, 2, 1235})).nodes();
+        REQUIRE(mid_q->nodes_get_list(&(n2)) == 0);
+    }
+
+    SECTION("Set and retrieve a single way")
+    {
+        osmid_t way_id = 1;
+        double lat = 12.3456789;
```

Comment:
These can be `const`. Probably lots of other places in the code where `const` could be added.